File: contrib/ifunc/julia/result.jl
##
# Dynamic adaptive pointer chasing application
#
# Author: Valentin Churavy
# Massachussetts Institute of Technology, 2022
#
# Original implementation:
# Author: Luis E. P.
# Copyright (C) 2021-2022 Arm Ltd. All rights reserved.
#
##
using UCX

include("UCXAsap.jl")
import .UCXAsap

include("IFunc.jl")
import .IFunc 

include("unsafe.jl")
include("common.jl")

function bound(source_args::Ptr{Cvoid}, source_args_size::Csize_t)::Csize_t
	return source_args_size
end

function init(source_args::Ptr{Cvoid}, source_args_size::Csize_t, payload::Ptr{Cvoid}, payload_size::Csize_t)::Cint
    result_src = Base.unsafe_convert(Ptr{UInt64}, source_args)
    result_pay = Base.unsafe_convert(Ptr{UInt64}, payload)

    Base.unsafe_store!(result_pay, Base.unsafe_load(result_src))

    return Cint(0)
end

function main(payload::Ptr{Cvoid}, payload_size::Csize_t, target_args::Ptr{Cvoid})::Cvoid
    result_pay = Base.unsafe_convert(Ptr{UInt64}, payload)
    result_tgt = Base.unsafe_convert(Ptr{UInt64}, target_args)

    Base.unsafe_store!(result_tgt, Base.unsafe_load(result_pay))

    return nothing
end

const fns = (
    (bound, Tuple{Ptr{Cvoid}, Csize_t}),
    (init, Tuple{Ptr{Cvoid}, Csize_t, Ptr{Cvoid}, Csize_t}),
    (main, Tuple{Ptr{Cvoid}, Csize_t, Ptr{Cvoid}})
)

IFunc.Compiler.emit(fns, "jresult", "$(Sys.ARCH)-unknown-linux-gnu")

File: contrib/ifunc/result.c
/*
 * Stops the ifunc polling loop. TODO: pure?
 * Author: Luis E. P.
 * Copyright (C) 2021-2022 Arm Ltd. All rights reserved.
 *
 */

#include <stdint.h>
#include <stdbool.h>
#include <stdio.h>

#include <ucp/api/ucp.h>

size_t result_payload_bound(void *source_args, size_t source_args_size)
{
	(void)source_args;
	return source_args_size;
}


int result_payload_init(void *source_args,
						size_t source_args_size,
						void *payload,
						size_t *payload_size)
{
    (void)source_args_size;
    (void)payload_size;

    uint64_t *result_src = (uint64_t *)source_args;
    uint64_t *result_pay = (uint64_t *)payload;

    *result_pay = *result_src;

    return 0;
}

// TODO: maybe make a more ellaborate struct with validity and stuff
void result_main(void *payload, size_t payload_size, void *target_args)
{
	(void)payload_size;

	uint64_t *result_pay = (uint64_t *)payload;
	uint64_t *result_tgt = (uint64_t *)target_args;

	*result_tgt = *result_pay;
}
